src/cli/build.rs

```diff
@@ -21,53 +40,243 @@ pub struct BuildCommand {
         println!(
-            "Writing standalone binary to {}",
-            style(output_path_displayed).blue()
+            "   {} standalone binary to {}",
+            style("Write").blue().bold(),
+            style(output_path.display()).underlined()
         );
-        write_executable_file_to(output_path, patched_bin).await?;
+        write_file_to(output_path, patched_bin, 0o755).await?; // Read & execute for all, write for owner
 
         Ok(ExitCode::SUCCESS)
     }
 }
 
-async fn write_executable_file_to(path: impl AsRef<Path>, bytes: impl AsRef<[u8]>) -> Result<()> {
+/// Wrapper function to asynchronously create a file at the given path with the given contents and permissions
+async fn write_file_to(
+    path: impl AsRef<Path>,
+    bytes: impl AsRef<[u8]>,
+    perms: u32,
+) -> Result<File> {
     let mut options = fs::OpenOptions::new();
-    options.write(true).create(true).truncate(true);
+    options.write(true).read(true).create(true).truncate(true);
 
     #[cfg(unix)]
     {
-        options.mode(0o755); // Read & execute for all, write for owner
+        options.mode(perms);
     }
 
     let mut file = options.open(path).await?;
     file.write_all(bytes.as_ref()).await?;
 
-    Ok(())
+    Ok(file)
+}
+
+/// Possible ways in which the discovery and/or download of a base binary's path can error
+#[derive(Debug, Clone, Error, PartialEq)]
+pub enum BasePathDiscoveryError<T> {
+    /// An error in the decompression of the precompiled target
+    #[error("decompression error")]
+    Decompression(T),
+    #[error("precompiled base for target not found for {target}")]
+    TargetNotFound { target: String },
+    /// An error in the precompiled target download process
+    #[error("failed to download precompiled binary base")]
+    DownloadError(T),
+    /// An IO related error
+    #[error("a generic error related to an io operation occurred")]
+    IoError(T),
+    /// Safe to continue, the user did not request any cross-compilation
+    #[error("neither a custom base path or precompiled target name provided")]
+    None,
+}
+
+/// Discovers the path to the base executable to use for cross-compilation
+async fn get_base_exe_path(
+    base: Option<PathBuf>,
+    target: Option<String>,
+    output_path: &mut PathBuf,
+) -> Result<PathBuf> {
+    if let Some(base) = base {
+        output_path.set_extension(
+            base.extension()
+                .expect("failed to get extension of base binary"),
+        );
+
+        Ok(base)
+    } else if let Some(target_inner) = target {
+        let target_exe_extension = match target_inner.as_str() {
+            "windows-x86_64" => "exe",
+            _ => "bin",
+        };
+
+        let path = TARGET_BASE_DIR.join(format!("lune-{target_inner}.{target_exe_extension}"));
+
+        output_path.set_extension(if target_exe_extension == "bin" {
+            ""
+        } else {
+            target_exe_extension
+        });
```

Comment:
Since it is a binary, and not a CLI tool. CLI tools can technically have the extension, but they are usually not included for convenience sake. It just feels a bit correct to do so for a cached binary target.

Reply:
The downloaded file is the full Lune binary, including CLI. It seems less correct to me to try and signal that it's not just the plain Lune binary.

Reply:
True, but it is not intended to be executed normally, and the "bin" extension signifies that, regardless of the fact that it is a complete lune CLI. 

In the future, it would be better to use precompiled binaries with the CLI feature disabled for compiling standalone binaries instead.

Reply:
I don't really see us providing Lune with CLI features disabled unless we also distribute using something other than GitHub releases. If this changes in the future we can revisit

Reply:
Addressed in https://github.com/lune-org/lune/pull/162/commits/4f80e98b8a99d0cc4f676a80b89a6890df75b570.